pkg/vm/stackitem/json.go

```diff
@@ -0,0 +1,202 @@
+package stackitem
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	gio "io"
+	"math"
+	"math/big"
+
+	"github.com/nspcc-dev/neo-go/pkg/io"
+)
+
+// decoder is a wrapper around json.Decoder helping to mimic C# json decoder behaviour.
+type decoder struct {
+	json.Decoder
+
+	depth int
+}
+
+// MaxAllowedInteger is the maximum integer allowed to be encoded.
+const MaxAllowedInteger = 2<<53 - 1
+
+// maxJSONDepth is a maximum allowed depth-level of decoded JSON.
+const maxJSONDepth = 10
+
+// ToJSON encodes Item to JSON.
+// It behaves as following:
+//   ByteArray -> base64 string
+//   BigInteger -> number
+//   Bool -> bool
+//   Null -> null
+//   Array -> array
+//   Map -> map with keys as UTF-8 bytes
+func ToJSON(item Item) ([]byte, error) {
+	buf := io.NewBufBinWriter()
+	toJSON(buf, item)
+	if buf.Err != nil {
+		return nil, buf.Err
+	}
+	buf.Len()
+	return buf.Bytes(), nil
+}
+
+func toJSON(buf *io.BufBinWriter, item Item) {
```

Comment:
That's weird. `w.WriteB('[')`. `w.WriteB(',')`. `w.WriteB('}')`...

Is it possible to create some surrogate type and use the default `json.Marshaler`?

Reply:
The problem is that we should somehow handle recursive structures. C# implementation restricts it implicitly by checking the size after every sub-item is marshaled.
When using `MarshalJSON` this requires sharing some state across multiple calls, which is rather dirty IMO, as I expect `MarshalJSON` to be independent of context.
We can hide all `'['` in constants to make it more readable.

Reply:
Something like
```
type X struct {
    context marshallContext
    item StackItem
}
```
In the marshaller you check the context and either marshal or not the item.

Is the main problem that we have is the length check?

Reply:
I have added `MarshalJSON` implementation, to me it looks not much better and uses a bunch of additional allocations.

Reply:
2 main problems are length check and order of map items (which should nonetheless be a JSON object).

Reply:
Yeah, it's around the same in terms of code size and using additional allocations, OK, if it doesn't really simplify anything than it's not worth going that route.